### ISSUE TITLE
GPU Option slower than cpu on m1

### ISSUE THREAD

**Issue by id4thomas** [apple/tensorflow_macos#128]
`import tensorflow as tf
from tensorflow import keras

from tensorflow.python.framework.ops import disable_eager_execution
disable_eager_execution() 

from tensorflow.python.compiler.mlcompute import mlcompute
mlcompute.set_mlc_device(device_name='gpu')

import numpy as np
print(tf.__version__)
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat',
               'Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']

train_images = train_images / 255.0

test_images = test_images / 255.0

model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(2000, activation='relu'),
    keras.layers.Dense(2000, activation='relu'),
    keras.layers.Dense(10, activation='softmax')
])
model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'],run_eagerly=False)

model.fit(train_images, train_labels, epochs=5)`

https://www.tensorflow.org/tutorials/keras/classification

Above is a code provided by TensorFlow for a simple DNN. When setting the compute device to 'gpu' system monitor shows gpu usage of about 92% but is much slower than using 'any'.

Epoch using 'any'
Epoch 1/5
60000/60000 [==============================] - 12s 201us/sample - loss: 0.4744 - accuracy: 0.8297

Epoch using 'gpu'
Epoch 1/5
60000/60000 [==============================] - 53s 882us/sample - loss: 0.4756 - accuracy: 0.8258

Tested with Mac mini with m1 16gb ram

**Comment by neurolabusc** [comment]
Is it possible that the any `option` to use the Apple Matrix Co-Processor (AMX) while the GPU path is restricted to Metal? This simple [demo](https://github.com/neurolabusc/Metal) shows the matrix multiplication is faster using the Accelerate framework relative to the GPU-based MPSMatrixMultiplication. As memory is shared, optimal performance might leverage dedicated hardware when available. My [benchmarks](https://github.com/neurolabusc/AppleSiliconForNeuroimaging) suggest the Accelerate framework is very fast. Unfortunately, unless Apple's Developer Ecosystem Engineering team helps openBLAS support these undocumented instructions users of popular tools like SciPy and NumPy will be unable to realize this potential.

**Comment by exgphe** [comment]
I think 'any' option uses CPU, GPU and neural engine altogether, so it is expected to be faster than GPU-only

**Comment by vashat** [comment]
> I think 'any' option uses CPU, GPU and neural engine altogether, so it is expected to be faster than GPU-only

No, neural engine is never used, it just sits there. When testing performance between 'cpu' and 'any' there is no difference at all. It seams ML Compute is not meant to use it.

**Comment by neurolabusc** [comment]
@vashat do the `cpu` and `any` use [AMX2](https://nod.ai/comparing-apple-m1-with-amx2-m1-with-neon/) whereas the `gpu` option does not? The Accelerate frameworks use of AMX2 allows it to be [faster](https://github.com/neurolabusc/Metal) than the GPU-based MPSMatrixMultiplication. This might explain the finding that `gpu` is slower than `any`.

**Comment by SharanSMenon** [comment]
Any updates here?

**Comment by guidiandrea** [comment]
I can confirm that I'm experiencing the same issue. When using 'gpu', it takes 1 minute per epoch on average, whereas if set on 'cpu' 12 seconds is the max (always per epoch).

**Comment by vashat** [comment]
Same here, on all examples I have run, cpu is much faster than gpu.

**Comment by id4thomas** [comment]
0.2a still has same issues.
Epoch 1/5
60000/60000 [==============================] - 48s 805us/sample - loss: 0.4770 - accuracy: 0.8253

Above is Train time running in Rosetta + Tensorflow Master branch built from source 
(https://github.com/tensorflow/tensorflow/issues/45434)
System monitor shows close to 500% CPU Utilization when training
Seems like cpu is still faster than gpu without m1 optimizations

**Comment by huang-ying** [comment]
Same problem here, Mac mini with 16G RAM, cpu is 7 to 10 times faster than gpu in fashion-mnist tutorial.

**Comment by nicomerli** [comment]
This is still an issue. Got 3x better performance on CPU. I'm running tests on a basic Seq2Seq model using Keras.

**Comment by dvcrn** [comment]
Same here. Anyone managed to find a workaround?

**Comment by willgryan** [comment]
Has there been any updates on this?

**Comment by xros** [comment]
Still having this problem. using GPU is slower than CPU, 1/4 performance
